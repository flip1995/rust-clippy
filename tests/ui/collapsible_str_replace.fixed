// run-rustfix

#![warn(clippy::collapsible_str_replace)]

fn get_filter() -> &'static str {
    "u"
}

fn main() {
    let misspelled = "hesuo worpd";

    let p = 'p';
    let s = 's';
    let u = 'u';
    let l = "l";

    // LINT CASES
    let _ = misspelled.replace(|c| matches!(c, 'u' | 's'), "l");

    let _ = misspelled.replace(|c| matches!(c, 'u' | 's'), l);

    let _ = misspelled.replace(|c| matches!(c, 'p' | 'u' | 's'), "l");

    let _ = misspelled
        .replace(|c| matches!(c, 'd' | 'p' | 'u' | 's'), "l");

    // FALLBACK CASES
    // If there are consecutive calls to `str::replace` and all or any chars are variables,
    // recommend the fallback `misspelled.replace(&[s, u, p], "l")`
    let _ = misspelled.replace(&['u' , s], "l");

    let _ = misspelled.replace(&['p' , 'u' , s], "l");

    let _ = misspelled.replace(&['p' , u , s], "l");

    let _ = misspelled.replace(&[p , u , s], "l");

    // NO LINT CASES
    let _ = misspelled.replace('s', "l");

    let _ = misspelled.replace(s, "l");

    // If the consecutive `str::replace` calls have different `to` arguments, do not lint
    let _ = misspelled.replace('s', "l").replace('u', "p");

    let _ = misspelled.replace(&get_filter(), "l");

    let _ = misspelled.replace(&['s', 'u', 'p'], "l");

    let _ = misspelled.replace(&['s', 'u', 'p'], l);

    let _ = misspelled.replace(&['s', 'u'], "l").replace(&['u', 'p'], "l");

    let _ = misspelled.replace('s', "l").replace(&['u', 'p'], "l");

    let _ = misspelled.replace(&['s', 'u'], "l").replace('p', "l");

    let _ = misspelled.replace(&['s', u, 'p'], "l");

    let _ = misspelled.replace(&[s, u, 'p'], "l");

    let _ = misspelled.replace(&[s, u, p], "l");

    let _ = misspelled.replace(&[s, u], "l").replace(&[u, p], "l");
}
