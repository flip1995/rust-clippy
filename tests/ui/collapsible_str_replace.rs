#![warn(clippy::collapsible_str_replace)]

fn get_filter() -> &'static str {
    "u"
}

fn main() {
    let misspelled = "hesuo worpd";

    let p = 'p';
    let s = 's';
    let u = 'u';

    // LINT CASES
    // If the first argument to a single `str::replace` call is a slice and none of the chars
    // are variables, recommend `collapsible_str_replace`
    let replacement = misspelled.replace(&['s', 'u', 'p'], "l");
    println!("{replacement}");

    // If there are consecutive calls to `str::replace` and none of the chars are variables,
    // recommend `collapsible_str_replace`
    let replacement = misspelled.replace('s', "l").replace('u', "l");
    println!("{replacement}");

    let replacement = misspelled.replace('s', "l").replace('u', "l").replace('p', "l");
    println!("{replacement}");

    // NO LINT CASES
    // If there is a single call to `str::replace` and the first argument is a char or a variable,
    // do not recommend `collapsible_str_replace`
    let replacement = misspelled.replace('s', "l");
    println!("{replacement}");

    let replacement = misspelled.replace(s, "l");
    println!("{replacement}");

    // If the `from` argument is of kind other than a slice or a char, do not lint
    let replacement = misspelled.replace(&get_filter(), "l");

    // NO LINT TIL IMPROVEMENT
    // If multiple `str::replace` calls contain slices and none of the chars are variables,
    // the first iteration does not recommend `collapsible_str_replace`
    let replacement = misspelled.replace(&['s', 'u', 'p'], "l").replace(&['s', 'p'], "l");
    println!("{replacement}");

    // If a mixture of `str::replace` calls with slice and char arguments are used for `from` arg,
    // the first iteration does not recommend `collapsible_str_replace`
    let replacement = misspelled.replace(&['s', 'u'], "l").replace('p', "l");
    println!("replacement");

    let replacement = misspelled.replace('p', "l").replace(&['s', 'u'], "l");

    // The first iteration of `collapsible_str_replace` will not create lint if the first argument to
    // a single `str::replace` call is a slice and one or more of its chars are variables
    let replacement = misspelled.replace(&['s', u, 'p'], "l");
    println!("{replacement}");

    let replacement = misspelled.replace(&[s, u, 'p'], "l");
    println!("{replacement}");

    let replacement = misspelled.replace(&[s, u, p], "l");
    println!("{replacement}");

    let replacement = misspelled.replace(&[s, u], "l").replace(&[u, p], "l");
    println!("{replacement}");

    // FALLBACK CASES
    // If there are consecutive calls to `str::replace` and all or any chars are variables,
    // recommend the fallback `misspelled.replace(&[s, u, p], "l")`
    let replacement = misspelled.replace(s, "l").replace('u', "l");
    println!("{replacement}");

    let replacement = misspelled.replace(s, "l").replace('u', "l").replace('p', "l");
    println!("{replacement}");

    let replacement = misspelled.replace(s, "l").replace(u, "l").replace('p', "l");
    println!("{replacement}");

    let replacement = misspelled.replace(s, "l").replace(u, "l").replace(p, "l");
    println!("{replacement}");
}
