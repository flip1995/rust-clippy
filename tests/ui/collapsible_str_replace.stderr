error: used consecutive `str::replace` call
  --> $DIR/collapsible_str_replace.rs:18:24
   |
LL |     let _ = misspelled.replace('s', "l").replace('u', "l");
   |                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: replace with: `replace(|c| matches!(c, 'u' | 's'), "l")`
   |
   = note: `-D clippy::collapsible-str-replace` implied by `-D warnings`

error: used consecutive `str::replace` call
  --> $DIR/collapsible_str_replace.rs:20:24
   |
LL |     let _ = misspelled.replace('s', l).replace('u', l);
   |                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: replace with: `replace(|c| matches!(c, 'u' | 's'), l)`

error: used consecutive `str::replace` call
  --> $DIR/collapsible_str_replace.rs:22:24
   |
LL |     let _ = misspelled.replace('s', "l").replace('u', "l").replace('p', "l");
   |                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: replace with: `replace(|c| matches!(c, 'p' | 'u' | 's'), "l")`

error: used consecutive `str::replace` call
  --> $DIR/collapsible_str_replace.rs:25:10
   |
LL |           .replace('s', "l")
   |  __________^
LL | |         .replace('u', "l")
LL | |         .replace('p', "l")
LL | |         .replace('d', "l");
   | |__________________________^ help: replace with: `replace(|c| matches!(c, 'd' | 'p' | 'u' | 's'), "l")`

error: used consecutive `str::replace` call
  --> $DIR/collapsible_str_replace.rs:33:24
   |
LL |     let _ = misspelled.replace(s, "l").replace('u', "l");
   |                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: replace with: `replace(&['u' , s], "l")`

error: used consecutive `str::replace` call
  --> $DIR/collapsible_str_replace.rs:35:24
   |
LL |     let _ = misspelled.replace(s, "l").replace('u', "l").replace('p', "l");
   |                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: replace with: `replace(&['p' , 'u' , s], "l")`

error: used consecutive `str::replace` call
  --> $DIR/collapsible_str_replace.rs:37:24
   |
LL |     let _ = misspelled.replace(s, "l").replace(u, "l").replace('p', "l");
   |                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: replace with: `replace(&['p' , u , s], "l")`

error: used consecutive `str::replace` call
  --> $DIR/collapsible_str_replace.rs:39:24
   |
LL |     let _ = misspelled.replace(s, "l").replace(u, "l").replace(p, "l");
   |                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: replace with: `replace(&[p , u , s], "l")`

error: aborting due to 8 previous errors

